pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConnectionHandler.java

```diff
@@ -41,10 +41,13 @@ public class ConnectionHandler {
     // Start with -1L because it gets incremented before sending on the first connection
     private volatile long epoch = -1L;
     protected volatile long lastConnectionClosedTimestamp = 0L;
+    private volatile CompletableFuture<Void> connectFuture;
```

Comment:
Look `connectFuture` is for controlling the state of `grabCnx `, could it only use `boolean`?

Reply:
Good suggestion. It's addressed.